vmm/src/cpu.rs

```diff
@@ -1061,7 +1105,44 @@ impl CpuManager {
                             return;
                         }
                     }
+
+                    #[cfg(not(feature = "kvm"))]
                     extern "C" fn handle_signal(_: i32, _: *mut siginfo_t, _: *mut c_void) {}
+                    #[cfg(feature = "kvm")]
+                    extern "C" fn handle_signal(_: i32, _: *mut siginfo_t, _: *mut c_void) {
+                        if IS_IN_SHUTDOWN.load(Ordering::SeqCst) {
+                            let mut lock = STATIC_VCPUS.write().unwrap();
+                            // TODO We should also revert the mmap()
+                            lock.drain(..);
+
+                            // thread::current() panics in shutdown phase, when threadlocal data was
+                            // already destroyed. We should instead prevent delivering signals in
+                            // that step.
+                            return;
+                        }
+
+                        let thread = self::thread::current();
+                        let name = thread.name().unwrap();
+                        let prefix = "vcpu";
+                        let id = &name[prefix.len()..];
+                        let vcpu_id = id.parse::<u8>().unwrap();
+                        let vcpus = STATIC_VCPUS.read().unwrap();
+                        for (vcpu_id_, kvm_run) in &*vcpus {
+                            let kvm_run = kvm_run.0;
+                            if  *vcpu_id_ == vcpu_id {
+                                /*let kvm_run = unsafe {  kvm_run.as_mut() }.unwrap();
+                                kvm_run.immediate_exit = std::hint::black_box(1);*/
+
+                                // Write the second field: immediate exit
+                                let ptr_kvm_run_immediate_exit = unsafe { kvm_run.cast::<u8>().add(1) };
+                                unsafe {
+                                    core::ptr::write_volatile(ptr_kvm_run_immediate_exit, 1);
+                                }
+                            }
+                        }
```

Comment:
If we use a thread local for kvm_run as I suggested above then this would instead be:

```suggestion
                        let kvm_run = KVM_RUN.get();
                        // Write the second field: immediate exit
                        let ptr_kvm_run_immediate_exit = unsafe { kvm_run.cast::<u8>().add(1) };
                        unsafe {
                            core::ptr::write_volatile(ptr_kvm_run_immediate_exit, 1);
                        }
```